### ISSUE TITLE
Int64::hash cause the time complexity for hash map insertion to degrade

### ISSUE THREAD

**Issue by tonyfettes** [moonbitlang/core#409]
bench:
```rust
fn main {
  let map = @hashmap.HashMap::new()
  for i = 0; i < 1000; i = i + 1 {
    for j = 0; j < 1000; j = j + 1 {
      let k = i.to_int64().lsl(32).lor(j.to_int64())
      map.set(k, (i + j, i * j))
    }
  }
}
```

I modify the output of the JavaScript output of the MoonBit program above:

```diff
 (function () {
   const map = moonbitlang$core$hashmap$$HashMap$new$1$(moonbitlang$core$hashmap$$HashMap$new$46$hasher$46$default$1$());
   let _tmp$14 = 0;
+  let start = Date.now();
   while (true) {
     const i = _tmp$14;
     if (i < 1000) {
       let _tmp$15 = 0;
+      console.log(`elapsed: ${Date.now() - start} ms`);
+      start = Date.now();
       while (true) {
         const j = _tmp$15;
         if (j < 1000) {
           const k = BigInt.asIntN(64, $i64_extend_i32_s(i) << (BigInt(32) & 63n)) | $i64_extend_i32_s(j);
           moonbitlang$core$hashmap$$HashMap$set$1$(map, k, { _0: i + j | 0, _1: Math.imul(i, j) | 0 });
           _tmp$15 = j + 1 | 0;
           continue;
         } else {
           break;
         }
       }
       _tmp$14 = i + 1 | 0;
       continue;
     } else {
       return;
     }
   }
 }());
```

<details>
<summary>Output</summary>

```plaintext
elapsed: 0 ms
elapsed: 1 ms
elapsed: 15 ms
elapsed: 15 ms
elapsed: 42 ms
elapsed: 12 ms
elapsed: 18 ms
elapsed: 127 ms
elapsed: 20 ms
elapsed: 23 ms
elapsed: 27 ms
elapsed: 30 ms
elapsed: 35 ms
elapsed: 37 ms
elapsed: 492 ms
elapsed: 48 ms
elapsed: 51 ms
elapsed: 57 ms
elapsed: 58 ms
elapsed: 61 ms
elapsed: 65 ms
elapsed: 67 ms
elapsed: 71 ms
elapsed: 75 ms
elapsed: 77 ms
elapsed: 81 ms
elapsed: 85 ms
elapsed: 2034 ms
elapsed: 96 ms
elapsed: 97 ms
elapsed: 98 ms
elapsed: 101 ms
elapsed: 104 ms
elapsed: 107 ms
elapsed: 112 ms
elapsed: 115 ms
elapsed: 121 ms
elapsed: 123 ms
elapsed: 127 ms
elapsed: 133 ms
elapsed: 136 ms
elapsed: 136 ms
elapsed: 143 ms
elapsed: 146 ms
elapsed: 147 ms
elapsed: 150 ms
elapsed: 154 ms
elapsed: 167 ms
elapsed: 162 ms
elapsed: 165 ms
elapsed: 172 ms
elapsed: 177 ms
elapsed: 176 ms
elapsed: 182 ms
elapsed: 3265 ms
elapsed: 195 ms
elapsed: 197 ms
elapsed: 199 ms
elapsed: 202 ms
elapsed: 208 ms
elapsed: 208 ms
elapsed: 213 ms
elapsed: 214 ms
elapsed: 218 ms
elapsed: 221 ms
elapsed: 226 ms
elapsed: 232 ms
elapsed: 235 ms
elapsed: 237 ms
elapsed: 242 ms
elapsed: 248 ms
elapsed: 250 ms
elapsed: 256 ms
elapsed: 261 ms
elapsed: 263 ms
elapsed: 267 ms
elapsed: 271 ms
elapsed: 274 ms
elapsed: 288 ms
elapsed: 290 ms
elapsed: 293 ms
elapsed: 293 ms
elapsed: 295 ms
elapsed: 306 ms
elapsed: 312 ms
elapsed: 323 ms
elapsed: 327 ms
elapsed: 329 ms
elapsed: 336 ms
elapsed: 335 ms
elapsed: 334 ms
elapsed: 339 ms
elapsed: 375 ms
elapsed: 379 ms
elapsed: 384 ms
elapsed: 401 ms
elapsed: 395 ms
elapsed: 380 ms
elapsed: 401 ms
elapsed: 392 ms
elapsed: 375 ms
elapsed: 382 ms
elapsed: 382 ms
elapsed: 393 ms
elapsed: 396 ms
elapsed: 393 ms
elapsed: 397 ms
elapsed: 11537 ms
elapsed: 420 ms
elapsed: 433 ms
elapsed: 422 ms
elapsed: 428 ms
elapsed: 420 ms
elapsed: 420 ms
elapsed: 435 ms
elapsed: 455 ms
elapsed: 444 ms
elapsed: 450 ms
elapsed: 453 ms
elapsed: 455 ms
elapsed: 474 ms
elapsed: 466 ms
elapsed: 463 ms
elapsed: 466 ms
elapsed: 477 ms
elapsed: 481 ms
elapsed: 528 ms
elapsed: 553 ms
elapsed: 597 ms
elapsed: 531 ms
elapsed: 498 ms
elapsed: 522 ms
elapsed: 697 ms
elapsed: 604 ms
elapsed: 593 ms
elapsed: 580 ms
elapsed: 543 ms
elapsed: 522 ms
elapsed: 534 ms
elapsed: 533 ms
elapsed: 533 ms
elapsed: 535 ms
elapsed: 549 ms
elapsed: 572 ms
elapsed: 553 ms
elapsed: 555 ms
elapsed: 562 ms
elapsed: 564 ms
elapsed: 567 ms
elapsed: 575 ms
elapsed: 605 ms
elapsed: 677 ms
elapsed: 670 ms
elapsed: 723 ms
elapsed: 689 ms
elapsed: 628 ms
elapsed: 610 ms
elapsed: 607 ms
elapsed: 612 ms
elapsed: 612 ms
elapsed: 640 ms
elapsed: 645 ms
elapsed: 632 ms
elapsed: 645 ms
elapsed: 641 ms
elapsed: 660 ms
elapsed: 647 ms
elapsed: 666 ms
elapsed: 679 ms
elapsed: 685 ms
elapsed: 676 ms
elapsed: 682 ms
elapsed: 690 ms
elapsed: 691 ms
elapsed: 725 ms
elapsed: 735 ms
elapsed: 697 ms
elapsed: 702 ms
elapsed: 718 ms
elapsed: 730 ms
elapsed: 907 ms
elapsed: 1521 ms
elapsed: 1220 ms
elapsed: 1105 ms
elapsed: 1002 ms
elapsed: 957 ms
elapsed: 1092 ms
elapsed: 1001 ms
elapsed: 792 ms
elapsed: 794 ms
elapsed: 783 ms
elapsed: 790 ms
elapsed: 842 ms
elapsed: 806 ms
elapsed: 962 ms
elapsed: 957 ms
elapsed: 955 ms
elapsed: 845 ms
elapsed: 906 ms
elapsed: 964 ms
elapsed: 1072 ms
elapsed: 905 ms
elapsed: 904 ms
elapsed: 1159 ms
elapsed: 945 ms
elapsed: 903 ms
elapsed: 902 ms
elapsed: 995 ms
elapsed: 925 ms
elapsed: 875 ms
elapsed: 880 ms
elapsed: 1016 ms
elapsed: 991 ms
elapsed: 44409 ms
elapsed: 900 ms
elapsed: 927 ms
elapsed: 954 ms
elapsed: 938 ms
elapsed: 954 ms
elapsed: 941 ms
elapsed: 930 ms
elapsed: 942 ms
elapsed: 973 ms
elapsed: 964 ms
elapsed: 982 ms
elapsed: 959 ms
elapsed: 1064 ms
elapsed: 1006 ms
elapsed: 999 ms
elapsed: 1009 ms
elapsed: 1029 ms
elapsed: 1032 ms
elapsed: 1022 ms
elapsed: 1027 ms
elapsed: 1072 ms
elapsed: 1043 ms
elapsed: 1062 ms
elapsed: 1132 ms
elapsed: 1054 ms
elapsed: 1093 ms
elapsed: 1076 ms
elapsed: 1104 ms
elapsed: 1084 ms
elapsed: 1072 ms
elapsed: 1151 ms
elapsed: 1070 ms
elapsed: 1105 ms
elapsed: 1105 ms
elapsed: 1172 ms
elapsed: 1107 ms
elapsed: 1118 ms
elapsed: 1170 ms
elapsed: 1137 ms
elapsed: 1197 ms
elapsed: 1206 ms
elapsed: 1147 ms
elapsed: 1155 ms
elapsed: 1158 ms
elapsed: 1200 ms
elapsed: 1202 ms
elapsed: 1231 ms
elapsed: 1163 ms
elapsed: 1215 ms
elapsed: 1187 ms
elapsed: 1242 ms
elapsed: 1387 ms
elapsed: 1364 ms
elapsed: 1211 ms
elapsed: 1412 ms
elapsed: 1352 ms
elapsed: 1285 ms
elapsed: 1363 ms
elapsed: 1539 ms
elapsed: 1322 ms
elapsed: 1372 ms
elapsed: 1427 ms
elapsed: 1389 ms
elapsed: 1536 ms
elapsed: 1405 ms
elapsed: 1366 ms
elapsed: 1297 ms
elapsed: 1341 ms
elapsed: 1362 ms
elapsed: 1364 ms
elapsed: 1445 ms
elapsed: 1427 ms
elapsed: 1359 ms
elapsed: 1417 ms
elapsed: 1419 ms
elapsed: 1492 ms
elapsed: 1479 ms
elapsed: 1444 ms
elapsed: 1534 ms
elapsed: 1387 ms
elapsed: 1431 ms
elapsed: 1458 ms
elapsed: 1450 ms
elapsed: 1412 ms
elapsed: 1498 ms
elapsed: 1425 ms
elapsed: 1445 ms
elapsed: 1496 ms
elapsed: 1515 ms
elapsed: 1494 ms
elapsed: 1452 ms
elapsed: 1514 ms
elapsed: 1472 ms
elapsed: 1494 ms
elapsed: 1582 ms
elapsed: 1481 ms
elapsed: 1556 ms
elapsed: 1611 ms
elapsed: 1510 ms
elapsed: 1655 ms
elapsed: 1807 ms
elapsed: 1613 ms
elapsed: 1646 ms
elapsed: 1653 ms
elapsed: 1749 ms
elapsed: 1753 ms
elapsed: 1754 ms
elapsed: 1763 ms
elapsed: 1725 ms
elapsed: 1770 ms
elapsed: 1639 ms
elapsed: 1728 ms
elapsed: 1864 ms
```
</details>

Plot of the running time (x: # k insertion, y: ms)
![3484708d-32d7-4026-bc24-f15798b9f479](https://github.com/moonbitlang/core/assets/29998228/48811b17-f1b5-44f3-948f-67806e6a161d)

Plot with large value removed.
![a59a0cac-5189-4c28-8e91-c65f7edac3a0](https://github.com/moonbitlang/core/assets/29998228/397f565b-ee6c-4f4c-8964-fb9b26eba0ed)

Current hash map API disallows user to specify initial capacity.

**Comment by tonyfettes** [comment]
Additional experiment:

Modified `hashmap.mbt`
```rust
/// Create new hash map.
pub fn HashMap::new[K, V](~hasher : Option[(K) -> Int] = None, ~capacity : Int = default_init_capacity) -> HashMap[K, V] {
  {
    size: 0,
    capacity: capacity,
    growAt: calc_grow_threshold(capacity),
    hasher,
    entries: @array.new(capacity, fn() { Empty }),
  }
}
```

bench:
```rust
fn main {
  let map = HashMap::new(~capacity=1048576)
  for i = 0; i < 1000; i = i + 1 {
    for j = 0; j < 1000; j = j + 1 {
      let k = i.to_int64().lsl(32).lor(j.to_int64())
      map.set(k, (i + j, i * j))
    }
  }
}
```

plot:
![3a1b4d8c-39ca-4609-9ce0-91a754c1f914](https://github.com/moonbitlang/core/assets/29998228/ad4391a8-0a6d-4b59-82d5-1118ff9e8c9d)

looks like even polynomial?

**Comment by tonyfettes** [comment]
after some investigation, the hash function of Int64 will potentially create many collisions:

```rust
pub fn hash(self : Int64) -> Int {
  let lo = self.to_int()
  let hi = self.lsr(32).to_int()
  lo.lxor(hi)
}
```

Notice if lo and hi are range over a similar interval, then this hash cannot distinguish (i, j) and (j, i)

**Comment by tonyfettes** [comment]
Migitation: convert `Int64` to `String` (which use xxhash internally) shows a much better performance.

**Comment by bobzhang** [comment]
@tonyfettes can you try to improve the hash over int64 direclty, e.g, https://github.com/jungomi/xxhash-wasm

**Comment by tonyfettes** [comment]
solved by #414
